### PR TITLE
Add image classification examples

### LINKED ISSUE #5: Add the SqueezeNet example
The SqueezeNet is one of [the first wave models](https://github.com/webmachinelearning/webnn/blob/master/op_compatibility/first_wave_models.md).

### PR DESCRIPTION
**Pull request by Honry**
Supports MobileNet and SqueezeNet with both 'NCHW' and 'NHWC' layouts.

Use MobileNet V2, SqueezeNet 1.1 from ONNX and MobileNet V2,
SqueezeNet 1.0 from TFLite as model baselines.

Centralize some common resources from style_transfer to 'common' folder BTW.

Fixed #5, #6

### CODE REVIEW

**Comment by Honry** [comment]
@huningxin, PTAL, thanks!

**Comment by huningxin** [comment]
Super. Thanks @Honry ! Do you have a preview link available?

**Comment by Honry** [comment]
@huningxin, here's the preview: https://honry.github.io/webnn-samples/image_classification/

**Comment by huningxin** [comment]
Thanks for sharing the preview link, @Honry ! It looks beautiful.

Some high-level comments:
1. The weights downloading takes time, could the progress be displayed on the UI? so user knows the status.
2. When I pick another image, it seems to download and build the model again, even without changing the model. Could it just compute?
3. Similar behavior when I just switch to camera view. I suppose it could reuse the compiled graph.
4. There is a warning saying "Cannot use 'in' operator to search for 'dispose' in null" when I pick another image.

File: common/utils.js [hunk 1]
```diff
@@ -28,15 +28,82 @@ export async function buildConstantByNpy(builder, url) {
...
   }
+  // TODO(Wanming): This is a workaround to transpose 'ihwo' to 'hwio',
+  // and will be removed once 'ihwo' filterLayout is supported.
+  if (isDepthwise) {
+    const a = tf.tensor(typedArray, dimensions, type);
+    const b = tf.transpose(a, [1, 2, 0, 3]);
+    const buffer = await b.buffer();
+    dimensions = b.shape;
+    typedArray = buffer.values;
+    tf.dispose();
```
Comment (huningxin) [inline common/utils.js:49]: I understand you need to dispose both `a` and `b`. However, probably it's good to wrap these code section into [`tf.tidy`](https://js.tensorflow.org/api/latest/#tidy).

**Comment by Honry** [comment]
@huningxin, many thanks for your comments! I've addressed most of the comments except the first one.

> The weights downloading takes time, could the progress be displayed on the UI? so user knows the status.

Since the weights downloading are wrapped in the `load()` method, we don't know the progress "out of the box", one way I can imagine is to hack each weight's loading to track the progress, but this would break the style of `load()` method and is not friendly for user. So I would suggest to skip this feature and see if there's better solution in future.

File: common/utils.js [hunk 2]
```diff
@@ -28,15 +28,82 @@ export async function buildConstantByNpy(builder, url) {
...
+  let typedArray = new TypedArrayConstructor(sizeOfShape(dimensions));
   const dataView = new DataView(npArray.data.buffer);
   const littleEndian = npArray.byteOrder === '<';
   for (let i = 0; i < sizeOfShape(dimensions); ++i) {
     typedArray[i] = dataView[`get` + type[0].toUpperCase() + type.substr(1)](
         i * TypedArrayConstructor.BYTES_PER_ELEMENT, littleEndian);
   }
+  // TODO(Wanming): This is a workaround to transpose 'ihwo' to 'hwio',
+  // and will be removed once 'ihwo' filterLayout is supported.
+  if (isDepthwise) {
```
Comment (huningxin) [inline common/utils.js:43]: If it is a workaround for a specific sample, I'd prefer to move it into the sample. You can add a `transpose` op in the graph after create the constant.

File: common/utils.js [hunk 3]
```diff
@@ -28,15 +28,82 @@ export async function buildConstantByNpy(builder, url) {
...
+    typedArray = buffer.values;
+    tf.dispose();
+  }
   return builder.constant({type, dimensions}, typedArray);
 }
+
+/**
+ * This method is used to covert input element to tensor data.
+ * @param {Object} inputElement, an object of HTML [<img> | <video>] element.
+ * @param {!Object<string, *>} options, an object of options to process
```
Comment (huningxin) [inline common/utils.js:57]: could you specify which options are supported?

File: image_classification/main.js [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+let stream = null;
+let loadTime = 0;
+let buildTime = 0;
+let computeTime = 0;
+let layout = 'nchw';
+const nchwOptions = {
+  mean: [0.485, 0.456, 0.406],
+  std: [0.229, 0.224, 0.225],
+  norm: true,
+  nchwFlag: true,
```
Comment (huningxin) [inline image_classification/main.js:31]: `inputLayout`?

File: image_classification/mobilenet_nchw.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+import {buildConstantByNpy} from '../common/utils.js';
+
+// MobileNet V2 baseline model with nchw layout
```
Comment (huningxin) [inline image_classification/mobilenet_nchw.js:5]: remove "baseline", add "input" before layout.

File: image_classification/mobilenet_nchw.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+export class MobileNetNchw {
```
Comment (huningxin) [inline image_classification/mobilenet_nchw.js:6]: `MobileNetV2Nchw

File: image_classification/main.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+  labelUrl: './labels/labels1000.txt',
+  inputDimensions: [1, 3, 224, 224],
+};
```
Comment (huningxin) [inline image_classification/main.js:34]: I think the options should be part of model, please move it respectively.

File: image_classification/mobilenet_nchw.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    const weightsName = prefix + '_weight.npy';
+    const weights =
+        await buildConstantByNpy(this.builder_, weightsName);
+    const biasName = prefix + '_bias.npy';
+    const bias =
+        await buildConstantByNpy(this.builder_, biasName);
+    const conv = this.builder_.add(
+        this.builder_.conv2d(input, weights, options),
+        this.builder_.reshape(bias, [1, -1, 1, 1]));
+    if (shouldClip) {
```
Comment (huningxin) [inline image_classification/mobilenet_nchw.js:23]: please name it `relu6`, that is used by the original model, we just use `clamp` to implement `relu6`.

File: image_classification/mobilenet_nchw.js [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    const prefix = './weights/mobilenet_nchw/gemm_' + name;
+    const weightsName = prefix + '_weight.npy';
+    const weights = await buildConstantByNpy(this.builder_, weightsName);
+    const biasName = prefix + '_bias.npy';
+    const bias = await buildConstantByNpy(this.builder_, biasName);
+    const options = {c: bias, bTranspose: true};
+    return this.builder_.gemm(input, weights, options);
+  }
+
+  async buildFire_(
```
Comment (huningxin) [inline image_classification/mobilenet_nchw.js:45]: `Fire` is not used in MobileNetV2 (it's used for SqueezeNet). According to the [paper](https://arxiv.org/pdf/1801.04381.pdf), we should build `Bottleneck` block.  And there are two types of bottleneck block, as it shows in the diagram:

image

And I'd suggest to create a helper to build depthwise conv explicitly. The helper could create grouped conv with groups == input channels.

**Comment by huningxin** [comment]
> one way I can imagine is to hack each weight's loading to track the progress, but this would break the style of `load()` method and is not friendly for user. So I would suggest to skip this feature and see if there's better solution in future.

sounds good. I agree the friendly code is more important in the context of WebNN API example.

File: common/utils.js (hunk 2 above)
Comment (Honry) [reply common/utils.js:43]: The `transpose` op in WebNN only supports transpose a `MLOperand`, that's why I have to do this workaround in `buildConstantByNpy`.

Comment (Honry) [reply common/utils.js:43]: Good suggestion! It's my misunderstanding the `transpose` op, I thought it could not transpose a `Constant`.

**Comment by Honry** [comment]
@huningxin, thanks for your further comments, really helpful! PTAL again, thanks!

File: image_classification/mobilenet_nhwc.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
+'use strict';
+
+import {buildConstantByNpy} from '../common/utils.js';
+
+/* eslint max-len: ["error", {"code": 120}] */
+
+// MobileNet V2 model with 'nhwc' input layout
+export class MobileNetV2Nhwc {
```
Comment (huningxin) [inline image_classification/mobilenet_nhwc.js:8]: Please align to `MobileNetV2Nchw` implementation.

File: image_classification/mobilenet_nhwc.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    const filterLayout = 'ohwi';
+    const input = this.builder_.input(
+        'input', {type: 'float32', dimensions: this.inputOptions.inputDimensions});
+    const conv0 = await this.buildConv_(
+        input, '90', 'Conv_Conv2D', true, {strides, autoPad, filterLayout});
+    const conv1 = await this.buildConv_(
+        conv0, '238', 'expanded_conv_depthwise_depthwise', true, {autoPad, groups: 32, filterLayout: 'hwio'});
+    const conv2 = await this.buildConv_(
+        conv1, '167', 'expanded_conv_project_Conv2D', false, {autoPad, filterLayout});
+    const fire0 = await this.buildBottleneck_(conv2, ['165', '99', '73'], '1', {strides, groups: 96});
```
Comment (huningxin) [inline image_classification/mobilenet_nhwc.js:78]: rename `fire0` to `bottleneck0`

File: image_classification/mobilenet_nhwc.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    const fire1 = await this.buildBottleneck_(fire0, ['3', '119', '115'], '2', {groups: 144});
+    const add0 = this.builder_.add(fire0, fire1);
```
Comment (huningxin) [inline image_classification/mobilenet_nhwc.js:80]: combine this `add` into `buildLinearBottleneck_` `shortcut` parameter.

**Comment by Honry** [comment]
@huningxin, thanks again! All comments fixed, PTAL, thanks!

File: common/.eslintrc.js [hunk 13]
```diff
@@ -3,5 +3,6 @@ module.exports = {
   globals: {
     'BigInt64Array': 'readonly',
     'BigUint64Array': 'readonly',
+    'tf': 'readonly',
```
Comment (huningxin) [inline common/.eslintrc.js:6]: do we still need this one?

**Review by huningxin — APPROVED** [review]
Nice work. LGTM with a nit. Please fix it and then we can merge. Thanks!

File: common/.eslintrc.js (hunk 13 above)
Comment (Honry) [reply common/.eslintrc.js]: Thanks, I've removed it.

**Comment by huningxin** [comment]
@Honry @BruceDai , there are many weight files (*.npy) in this PR. As the webnn-samples and tests in webnn-polyfill (probably later webnn-native) will share them, we may need to find a place to host them separately. Probably a new repo under webmachinelearning, say github.com/webmachinelearning/test-data?

Any thoughts?

also cc @anssiko for comments.

**Comment by Honry** [comment]
@huningxin, good idea! Then we can make it as a git submodule for both webnn-samples and webnn-polyfill.

**Comment by BruceDai** [comment]
@huningxin Yes, adding new repo is a good suggestion. How about this name github.com/webmachinelearning/models_zoo?

**Comment by huningxin** [comment]
> How about this name github.com/webmachinelearning/models_zoo?

@BruceDai , thanks for the suggestion. However I suppose we will not host any models (like .onnx or .tflite files) there. And webnn-polyfill and webnn-native may share the test data other than model weights.

**Comment by BruceDai** [comment]
@huningxin Thanks. That sounds good.

**Comment by anssiko** [comment]
I support @huningxin’s proposal for a new repo `webmachinelearning/test-data` to host reusable .npy weight files and other test data.

@huningxin feel free to create the repo and please add an appropriate readme to this new repo to note projects that are expected to reuse these assets.

**Comment by huningxin** [comment]
> feel free to create the repo and please add an appropriate readme to this new repo to note projects that are expected to reuse these assets.

done: https://github.com/webmachinelearning/test-data

@Honry , could you please extract the npy files in this PR and include them in a new PR to test-data repo.

We may create folder like 'test-data/models/mobilenetv2_nchw/`. I would suggest to post a README file for each model that includes the info of the origin and license.

Once the weights of mobilenet and squeezenet are available in the test-data repo, please also update the model tests in polyfill to use them.

Further, we may need to move LeNet and NSNet2 weights to test-data.

**Comment by Honry** [comment]
@huningxin, sure, I will make it as a git submodule to webnn-samples and webnn-polyfill.

**Comment by Honry** [comment]
@huningxin, I've removed the weight files in this PR. PTAL, thanks!

**Comment by huningxin** [comment]
LGTM. Great work, thanks again @Honry !

**Pull request merged**